File: tests/test_liveness.test.js
import test from 'node:test';
import assert from 'node:assert';

/*
 * TEST SUITE: STORY 29.1.1: Resolve STATIC_APPLICATION
 * Defect ID: DF-LIVENESS-01 (Severity: HIGH)
 * Target: src/main.js
 *
 * Acceptance Criteria:
 * 1. Given the web application entrypoint src/main.js is mounted to document.getElementById('app'),
 *    When 2.5 seconds elapse in the browser, Then the application must execute a continuous
 *    render loop via requestAnimationFrame and reflect state mutations on the DOM or canvas.
 * 2. Given the active application entrypoint (src/main.js), When the browser renders the page,
 *    Then the entrypoint must bind all animation cycles and dynamic updates directly to the live DOM container.
 */

// --- Deterministic Mock Browser Environment ---

class MockCanvasRenderingContext2D {
  constructor(canvas) {
    this.canvas = canvas;
    this.drawHistory = [];
    this.fillStyle = '#000000';
    this.strokeStyle = '#000000';
    this.lineWidth = 1;
    this.font = '10px sans-serif';
  }

  _record(method, args) {
    this.drawHistory.push({
      method,
      args: structuredClone(args),
      timestamp: globalThis.performance.now(),
      fillStyle: this.fillStyle,
    });
  }

  clearRect(x, y, w, h) { this._record('clearRect', [x, y, w, h]); }
  fillRect(x, y, w, h) { this._record('fillRect', [x, y, w, h]); }
  strokeRect(x, y, w, h) { this._record('strokeRect', [x, y, w, h]); }
  fillText(text, x, y) { this._record('fillText', [text, x, y]); }
  strokeText(text, x, y) { this._record('strokeText', [text, x, y]); }
  beginPath() { this._record('beginPath', []); }
  closePath() { this._record('closePath', []); }
  stroke() { this._record('stroke', []); }
  fill() { this._record('fill', []); }
  arc(x, y, radius, startAngle, endAngle) { this._record('arc', [x, y, radius, startAngle, endAngle]); }
  drawImage(...args) { this._record('drawImage', args); }
}

class MockElement {
  constructor(tagName, id = '') {
    this.tagName = tagName.toUpperCase();
    this.id = id;
    this.children = [];
    this.parentElement = null;
    this.attributes = new Map();
    this._textContent = '';
    this._innerHTML = '';
    this.mutationEvents = [];
    this._ctx = null;
    this.width = 800;
    this.height = 600;
  }

  get textContent() {
    return this._textContent;
  }

  set textContent(val) {
    const prev = this._textContent;
    this._textContent = String(val);
    this.mutationEvents.push({
      type: 'textContent',
      from: prev,
      to: this._textContent,
      timestamp: globalThis.performance.now(),
    });
  }

  get innerHTML() {
    return this._innerHTML;
  }

  set innerHTML(val) {
    const prev = this._innerHTML;
    this._innerHTML = String(val);
    this.mutationEvents.push({
      type: 'innerHTML',
      from: prev,
      to: this._innerHTML,
      timestamp: globalThis.performance.now(),
    });
  }

  appendChild(child) {
    child.parentElement = this;
    this.children.push(child);
    this.mutationEvents.push({
      type: 'appendChild',
      target: child,
      timestamp: globalThis.performance.now(),
    });
    return child;
  }

  removeChild(child) {
    const idx = this.children.indexOf(child);
    if (idx !== -1) {
      this.children.splice(idx, 1);
      child.parentElement = null;
      this.mutationEvents.push({
        type: 'removeChild',
        target: child,
        timestamp: globalThis.performance.now(),
      });
    }
    return child;
  }

  contains(node) {
    if (!node) return false;
    if (node === this) return true;
    for (const child of this.children) {
      if (child.contains ? child.contains(node) : child === node) {
        return true;
      }
    }
    return false;
  }

  setAttribute(name, value) {
    this.attributes.set(name, String(value));
    this.mutationEvents.push({
      type: 'attribute',
      name,
      value: String(value),
      timestamp: globalThis.performance.now(),
    });
  }

  getAttribute(name) {
    return this.attributes.get(name) ?? null;
  }

  getContext(contextType) {
    if (this.tagName === 'CANVAS' && contextType === '2d') {
      if (!this._ctx) {
        this._ctx = new MockCanvasRenderingContext2D(this);
      }
      return this._ctx;
    }
    return null;
  }
}

class MockDocument {
  constructor() {
    this.elementsById = new Map();
    this.body = new MockElement('BODY');
  }

  getElementById(id) {
    return this.elementsById.get(id) || null;
  }

  createElement(tagName) {
    return new MockElement(tagName);
  }

  _registerElement(element) {
    if (element.id) {
      this.elementsById.set(element.id, element);
    }
  }
}

class BrowserEnvironmentSimulator {
  constructor() {
    this.currentTime = 0;
    this.rafIdCounter = 0;
    this.scheduledRafCallbacks = new Map();
    this.totalRafDispatched = 0;

    this.document = new MockDocument();
    this.appContainer = new MockElement('DIV', 'app');
    this.document._registerElement(this.appContainer);
    this.document.body.appendChild(this.appContainer);

    this.originalGlobals = {
      window: globalThis.window,
      document: globalThis.document,
      requestAnimationFrame: globalThis.requestAnimationFrame,
      cancelAnimationFrame: globalThis.cancelAnimationFrame,
      performance: globalThis.performance,
      HTMLCanvasElement: globalThis.HTMLCanvasElement,
    };
  }

  install() {
    globalThis.performance = {
      now: () => this.currentTime,
    };

    globalThis.requestAnimationFrame = (callback) => {
      const id = ++this.rafIdCounter;
      this.scheduledRafCallbacks.set(id, callback);
      return id;
    };

    globalThis.cancelAnimationFrame = (id) => {
      this.scheduledRafCallbacks.delete(id);
    };

    globalThis.document = this.document;
    globalThis.window = {
      document: this.document,
      requestAnimationFrame: globalThis.requestAnimationFrame,
      cancelAnimationFrame: globalThis.cancelAnimationFrame,
      performance: globalThis.performance,
    };
    globalThis.HTMLCanvasElement = MockElement;
  }

  restore() {
    globalThis.window = this.originalGlobals.window;
    globalThis.document = this.originalGlobals.document;
    globalThis.requestAnimationFrame = this.originalGlobals.requestAnimationFrame;
    globalThis.cancelAnimationFrame = this.originalGlobals.cancelAnimationFrame;
    globalThis.performance = this.originalGlobals.performance;
    globalThis.HTMLCanvasElement = this.originalGlobals.HTMLCanvasElement;
  }

  /**
   * Advances deterministic simulation time, triggering animation frames at discrete steps.
   * Standard 60Hz step ~16.666ms.
   */
  advanceTime(durationMs, stepMs = 16.666) {
    const targetTime = this.currentTime + durationMs;
    while (this.currentTime + stepMs <= targetTime) {
      this.currentTime += stepMs;
      const currentBatch = Array.from(this.scheduledRafCallbacks.entries());
      this.scheduledRafCallbacks.clear();

      for (const [id, callback] of currentBatch) {
        this.totalRafDispatched++;
        callback(this.currentTime);
      }
    }
    this.currentTime = targetTime;
  }
}

/**
 * Dynamically loads and bootstraps the entrypoint.
 * Evaluates both immediate auto-mount patterns and explicit exported hooks (mount/init).
 */
async function bootstrapEntrypoint(appContainer) {
  // Use unique query parameter to ensure fresh module instantiation per test
  const modulePath = `../src/main.js?test_run=${Date.now()}_${Math.random()}`;
  const entrypoint = await import(modulePath);

  if (typeof entrypoint.mount === 'function') {
    entrypoint.mount(appContainer);
  } else if (typeof entrypoint.init === 'function') {
    entrypoint.init();
  } else if (typeof entrypoint.default === 'function') {
    entrypoint.default();
  }

  return entrypoint;
}

// --- Unit Tests ---

test('DF-LIVENESS-01: Continuous render loop execution over 2.5s observation window', async (t) => {
  const env = new BrowserEnvironmentSimulator();
  env.install();

  try {
    await bootstrapEntrypoint(env.appContainer);

    // Initial frame request must be scheduled immediately upon mounting
    assert.ok(
      env.scheduledRafCallbacks.size > 0,
      'AC-1 Violation: Application mounted but did not request an initial animation frame via requestAnimationFrame.'
    );

    // Advance by exactly 2500ms (2.5 seconds observation window)
    const observationWindowMs = 2500;
    const stepMs = 16.666; // Standard 60 FPS frame delta
    const expectedMinFrames = Math.floor(observationWindowMs / stepMs) - 5; // Allow minor variance buffer

    env.advanceTime(observationWindowMs, stepMs);

    // Continuous execution verification
    assert.ok(
      env.totalRafDispatched >= expectedMinFrames,
      `AC-1 Violation (DF-LIVENESS-01): Render loop stalled or terminated prematurely. ` +
      `Expected at least ${expectedMinFrames} continuous frames over 2.5s, but only received ${env.totalRafDispatched}.`
    );

    // Ensure loop is self-sustaining and continuous (a next frame must be queued at t = 2500ms)
    assert.ok(
      env.scheduledRafCallbacks.size > 0,
      'AC-1 Violation: Render loop is not continuous; no subsequent animation frame was requested at the end of 2.5s.'
    );
  } finally {
    env.restore();
  }
});

test('DF-LIVENESS-01: State mutations reflected on DOM or Canvas across 2.5-second observation window', async (t) => {
  const env = new BrowserEnvironmentSimulator();
  env.install();

  try {
    await bootstrapEntrypoint(env.appContainer);

    // Locate active visual target inside #app (Canvas or dynamic Element)
    const findCanvas = (el) => {
      if (el.tagName === 'CANVAS') return el;
      for (const child of el.children) {
        const found = findCanvas(child);
        if (found) return found;
      }
      return null;
    };

    // Capture state snapshots across 5 discrete checkpoints during the 2.5s window:
    // t0 = 0ms, t1 = 500ms, t2 = 1000ms, t3 = 1500ms, t4 = 2000ms, t5 = 2500ms
    const checkpoints = [500, 500, 500, 500, 500]; // Incremental steps summing to 2500ms
    const snapshots = [];

    const captureSnapshot = (timestamp) => {
      const canvas = findCanvas(env.appContainer);
      const ctx = canvas ? canvas.getContext('2d') : null;
      return {
        timestamp,
        canvasCallsCount: ctx ? ctx.drawHistory.length : 0,
        lastDrawCall: ctx && ctx.drawHistory.length > 0 ? structuredClone(ctx.drawHistory.at(-1)) : null,
        domText: env.appContainer.textContent,
        domInnerHTML: env.appContainer.innerHTML,
        domMutationCount: env.appContainer.mutationEvents.length,
      };
    };

    snapshots.push(captureSnapshot(0));

    for (const step of checkpoints) {
      env.advanceTime(step);
      snapshots.push(captureSnapshot(env.currentTime));
    }

    assert.strictEqual(snapshots.length, 6, 'Internal Test Error: Expected 6 temporal state snapshots.');

    // Evaluation for Canvas-based mutations:
    const initialSnapshot = snapshots[0];
    const finalSnapshot = snapshots[snapshots.length - 1];

    let hasCanvasMutations = false;
    if (finalSnapshot.canvasCallsCount > initialSnapshot.canvasCallsCount) {
      // Validate that draw operations were not identical static duplicates
      const canvas = findCanvas(env.appContainer);
      const ctx = canvas.getContext('2d');
      const drawCallsOverTime = ctx.drawHistory;

      // Ensure draw calls occurred across the full window, including near t = 2500ms
      const lateFrames = drawCallsOverTime.filter((call) => call.timestamp >= 2000);
      assert.ok(
        lateFrames.length > 0,
        'AC-1 Violation (DF-LIVENESS-01): Canvas frame renders ceased before 2.5s observation window elapsed.'
      );

      // Verify that arguments or draw states mutated between initial and final snapshots
      const earlyArgs = JSON.stringify(drawCallsOverTime.slice(0, 5).map(c => ({ method: c.method, args: c.args })));
      const lateArgs = JSON.stringify(lateFrames.slice(-5).map(c => ({ method: c.method, args: c.args })));
      
      const statesDiverged = earlyArgs !== lateArgs;
      assert.ok(
        statesDiverged,
        'AC-1 Violation (DF-LIVENESS-01): Static painting detected. Canvas rendered frames without any state/coordinate mutations.'
      );
      hasCanvasMutations = true;
    }

    // Evaluation for DOM-based mutations:
    let hasDOMMutations = false;
    if (finalSnapshot.domText !== initialSnapshot.domText ||
        finalSnapshot.domInnerHTML !== initialSnapshot.domInnerHTML ||
        finalSnapshot.domMutationCount > initialSnapshot.domMutationCount) {
      hasDOMMutations = true;
    }

    // High Severity Defect Check: Neither DOM text nor canvas pixels underwent any state mutations
    assert.ok(
      hasCanvasMutations || hasDOMMutations,
      'DF-LIVENESS-01 Failure: Static application detected! Neither DOM text nor canvas pixels ' +
      'underwent any state mutations or dynamic frame renders over the 2.5-second observation window.'
    );
  } finally {
    env.restore();
  }
});

test('AC-2: Live DOM Container Binding and Structural Integrity', async (t) => {
  const env = new BrowserEnvironmentSimulator();
  env.install();

  try {
    await bootstrapEntrypoint(env.appContainer);

    // Verify that the entrypoint mounted dynamic visual nodes to document.getElementById('app')
    assert.ok(
      env.appContainer.children.length > 0 || env.appContainer.textContent.length > 0 || env.appContainer.innerHTML.length > 0,
      'AC-2 Violation: Entrypoint src/main.js failed to attach components to document.getElementById("app").'
    );

    // Advance 500ms to allow rendering updates to dispatch
    env.advanceTime(500);

    // Architectural Invariant: Never produce an isolated, unmounted file.
    // Verify all active canvas elements or mutating elements reside within the live #app container
    const isNodeInLiveApp = (node) => env.appContainer.contains(node);

    const canvasElements = [];
    const collectCanvases = (el) => {
      if (el.tagName === 'CANVAS') canvasElements.push(el);
      for (const child of el.children) collectCanvases(child);
    };
    collectCanvases(env.appContainer);

    for (const canvas of canvasElements) {
      assert.ok(
        isNodeInLiveApp(canvas),
        'Architectural Invariant Violation: Canvas element is detached from the active DOM container (#app).'
      );

      const ctx = canvas.getContext('2d');
      assert.ok(
        ctx !== null,
        'AC-2 Violation: Unable to acquire active 2D rendering context from mounted canvas.'
      );

      assert.ok(
        ctx.drawHistory.length > 0,
        'AC-2 Violation: Mounted canvas inside #app received zero render calls from the live animation loop.'
      );
    }
  } finally {
    env.restore();
  }
});

test('DF-LIVENESS-01: Render loop resilience against frame rate drops', async (t) => {
  const env = new BrowserEnvironmentSimulator();
  env.install();

  try {
    await bootstrapEntrypoint(env.appContainer);

    // Simulate unstable browser frame timing (e.g. jank, variable delta) across 2500ms
    const variableSteps = [33.3, 16.6, 50.0, 16.6, 100.0, 16.6, 16.6, 250.0];
    let elapsed = 0;
    let stepIndex = 0;

    while (elapsed < 2500) {
      const step = variableSteps[stepIndex % variableSteps.length];
      const remaining = 2500 - elapsed;
      const actualStep = Math.min(step, remaining);
      env.advanceTime(actualStep, actualStep);
      elapsed += actualStep;
      stepIndex++;
    }

    // Verify render loop recovers and maintains state mutation progression
    assert.ok(
      env.totalRafDispatched > 0,
      'AC-1 Violation: Render loop halted during variable frame pacing.'
    );

    assert.ok(
      env.scheduledRafCallbacks.size > 0,
      'AC-1 Violation: Render loop terminated when handling variable time intervals.'
    );
  } finally {
    env.restore();
  }
});

File: tests/test_liveness_test.py
import test, { describe, it, beforeEach, afterEach } from 'node:test';
import assert from 'node:assert/strict';

/**
 * STORY 53.1.1: Resolve STATIC_APPLICATION
 * Defect ID: DF-LIVENESS-01
 * 
 * Target: src/main.js
 * Verification of active animation loop, liveness over 2.5s observation window,
 * and continuous DOM/Canvas state mutation without user input.
 */

// --- Mock Browser & Virtual Clock Harness ---

class MockCanvasRenderingContext2D {
  constructor(canvas, recordMutation) {
    this.canvas = canvas;
    this.recordMutation = recordMutation;
    this._fillStyle = '#000000';
    this._strokeStyle = '#000000';
    this.renderCallCount = 0;
  }

  get fillStyle() {
    return this._fillStyle;
  }

  set fillStyle(val) {
    this._fillStyle = val;
  }

  get strokeStyle() {
    return this._strokeStyle;
  }

  set strokeStyle(val) {
    this._strokeStyle = val;
  }

  clearRect(x, y, w, h) {
    this.renderCallCount++;
    this.recordMutation('canvas:clearRect', { x, y, w, h, callCount: this.renderCallCount });
  }

  fillRect(x, y, w, h) {
    this.renderCallCount++;
    this.recordMutation('canvas:fillRect', { x, y, w, h, fillStyle: this._fillStyle, callCount: this.renderCallCount });
  }

  strokeRect(x, y, w, h) {
    this.renderCallCount++;
    this.recordMutation('canvas:strokeRect', { x, y, w, h, strokeStyle: this._strokeStyle, callCount: this.renderCallCount });
  }

  fillText(text, x, y) {
    this.renderCallCount++;
    this.recordMutation('canvas:fillText', { text, x, y, callCount: this.renderCallCount });
  }

  beginPath() {}
  closePath() {}
  stroke() {
    this.renderCallCount++;
    this.recordMutation('canvas:stroke', { callCount: this.renderCallCount });
  }
  fill() {
    this.renderCallCount++;
    this.recordMutation('canvas:fill', { callCount: this.renderCallCount });
  }
  arc() {}
  drawImage() {
    this.renderCallCount++;
    this.recordMutation('canvas:drawImage', { callCount: this.renderCallCount });
  }

  getImageData(sx, sy, sw, sh) {
    // Return mock pixel buffer reflecting current renderCallCount
    const data = new Uint8ClampedArray(sw * sh * 4);
    const val = this.renderCallCount % 256;
    data.fill(val);
    return { data, width: sw, height: sh };
  }
}

class MockElement {
  constructor(tagName, id = '', recordMutation) {
    this.tagName = tagName.toUpperCase();
    this.id = id;
    this.recordMutation = recordMutation;
    this.children = [];
    this._textContent = '';
    this._innerHTML = '';
    this.attributes = new Map();
    this._context2D = null;
    this.width = 800;
    this.height = 600;
  }

  get textContent() {
    return this._textContent;
  }

  set textContent(val) {
    const prev = this._textContent;
    this._textContent = String(val);
    if (prev !== this._textContent) {
      this.recordMutation('dom:textContent', { id: this.id, text: this._textContent });
    }
  }

  get innerHTML() {
    return this._innerHTML;
  }

  set innerHTML(val) {
    const prev = this._innerHTML;
    this._innerHTML = String(val);
    if (prev !== this._innerHTML) {
      this.recordMutation('dom:innerHTML', { id: this.id, html: this._innerHTML });
    }
  }

  setAttribute(name, value) {
    this.attributes.set(name, String(value));
    this.recordMutation('dom:setAttribute', { id: this.id, name, value });
  }

  getAttribute(name) {
    return this.attributes.get(name) || null;
  }

  appendChild(child) {
    this.children.push(child);
    this.recordMutation('dom:appendChild', { parentId: this.id, childTag: child.tagName, childId: child.id });
    return child;
  }

  getContext(type) {
    if (this.tagName === 'CANVAS' && type === '2d') {
      if (!this._context2D) {
        this._context2D = new MockCanvasRenderingContext2D(this, this.recordMutation);
      }
      return this._context2D;
    }
    return null;
  }

  addEventListener() {}
  removeEventListener() {}
}

class VirtualEnvironment {
  constructor() {
    this.currentTime = 0;
    this.mutationLog = [];
    this.rafCallbacks = new Map();
    this.timerCallbacks = new Map();
    this.nextRafId = 1;
    this.nextTimerId = 1;
    this.elements = new Map();

    const appContainer = new MockElement('div', 'app', (type, meta) => this.logMutation(type, meta));
    this.elements.set('app', appContainer);

    this.setupGlobals();
  }

  logMutation(type, meta) {
    this.mutationLog.push({
      time: this.currentTime,
      type,
      meta,
    });
  }

  setupGlobals() {
    globalThis.window = globalThis;

    globalThis.document = {
      getElementById: (id) => {
        if (!this.elements.has(id)) {
          const el = new MockElement('div', id, (t, m) => this.logMutation(t, m));
          this.elements.set(id, el);
        }
        return this.elements.get(id);
      },
      createElement: (tagName) => {
        return new MockElement(tagName, '', (t, m) => this.logMutation(t, m));
      },
      body: new MockElement('body', 'body', (t, m) => this.logMutation(t, m)),
    };

    globalThis.requestAnimationFrame = (cb) => {
      const id = this.nextRafId++;
      this.rafCallbacks.set(id, cb);
      return id;
    };

    globalThis.cancelAnimationFrame = (id) => {
      this.rafCallbacks.delete(id);
    };

    globalThis.setInterval = (cb, delay) => {
      const id = this.nextTimerId++;
      this.timerCallbacks.set(id, { cb, delay: Math.max(delay || 0, 1), nextRun: this.currentTime + Math.max(delay || 0, 1), recurring: true });
      return id;
    };

    globalThis.clearInterval = (id) => {
      this.timerCallbacks.delete(id);
    };

    globalThis.setTimeout = (cb, delay) => {
      const id = this.nextTimerId++;
      this.timerCallbacks.set(id, { cb, delay: Math.max(delay || 0, 1), nextRun: this.currentTime + Math.max(delay || 0, 1), recurring: false });
      return id;
    };

    globalThis.clearTimeout = (id) => {
      this.timerCallbacks.delete(id);
    };

    globalThis.performance = {
      now: () => this.currentTime,
    };
  }

  advanceTime(ms, stepMs = 16.67) {
    const targetTime = this.currentTime + ms;
    while (this.currentTime < targetTime) {
      this.currentTime = Math.min(this.currentTime + stepMs, targetTime);

      // Execute queued requestAnimationFrame callbacks for this frame
      if (this.rafCallbacks.size > 0) {
        const currentCallbacks = Array.from(this.rafCallbacks.entries());
        this.rafCallbacks.clear();
        for (const [id, cb] of currentCallbacks) {
          cb(this.currentTime);
        }
      }

      // Execute due timer callbacks
      for (const [id, timer] of Array.from(this.timerCallbacks.entries())) {
        if (this.currentTime >= timer.nextRun) {
          timer.cb();
          if (timer.recurring) {
            timer.nextRun = this.currentTime + timer.delay;
          } else {
            this.timerCallbacks.delete(id);
          }
        }
      }
    }
  }

  cleanup() {
    delete globalThis.window;
    delete globalThis.document;
    delete globalThis.requestAnimationFrame;
    delete globalThis.cancelAnimationFrame;
    delete globalThis.setInterval;
    delete globalThis.clearInterval;
    delete globalThis.setTimeout;
    delete globalThis.clearTimeout;
    delete globalThis.performance;
  }
}

// Helper to initialize entrypoint supporting both default/named exports and top-level side effects
async function loadMainEntrypoint() {
  const cacheBust = `?t=${Date.now()}_${Math.random()}`;
  const module = await import(`../src/main.js${cacheBust}`);
  const appElement = globalThis.document.getElementById('app');

  if (typeof module.mount === 'function') {
    await module.mount(appElement);
  } else if (typeof module.init === 'function') {
    await module.init(appElement);
  } else if (typeof module.default === 'function') {
    await module.default(appElement);
  }

  return { module, appElement };
}

describe('STORY 53.1.1: Resolve STATIC_APPLICATION (Defect DF-LIVENESS-01)', () => {
  let env;

  beforeEach(() => {
    env = new VirtualEnvironment();
  });

  afterEach(() => {
    env.cleanup();
  });

  it('AC-1: mounts to document.getElementById("app") and starts continuous loop mutating DOM or Canvas at least once every 1000ms', async () => {
    await loadMainEntrypoint();

    const app = env.elements.get('app');
    assert.ok(
      app.children.length > 0 || app.textContent.length > 0 || app.innerHTML.length > 0,
      'App container document.getElementById("app") must be populated upon mounting src/main.js'
    );

    // Initial state after mount
    const initialLogCount = env.mutationLog.length;

    // Advance 1000ms in discrete frame steps
    env.advanceTime(1000);

    const mutationsInFirstSec = env.mutationLog.slice(initialLogCount);
    assert.ok(
      mutationsInFirstSec.length >= 1,
      `Expected at least 1 DOM or Canvas render mutation in the first 1000ms, but recorded ${mutationsInFirstSec.length}`
    );

    // Advance another 1000ms and verify continuous execution
    const beforeSec2 = env.mutationLog.length;
    env.advanceTime(1000);
    const mutationsInSec2 = env.mutationLog.slice(beforeSec2);

    assert.ok(
      mutationsInSec2.length >= 1,
      `Animation loop must continuously run; expected at least 1 mutation in the second 1000ms window, got ${mutationsInSec2.length}`
    );
  });

  it('AC-1: maximum elapsed interval between consecutive frame renders or DOM mutations must not exceed 1000ms', async () => {
    await loadMainEntrypoint();

    // Advance through 3000ms to observe continuous intervals
    const observationMs = 3000;
    env.advanceTime(observationMs);

    assert.ok(
      env.mutationLog.length >= 3,
      `Over a ${observationMs}ms execution, expected at least 3 mutations (1 every 1000ms), but got ${env.mutationLog.length}`
    );

    // Check interval between consecutive render timestamps
    let previousTimestamp = 0;
    for (let i = 0; i < env.mutationLog.length; i++) {
      const currentTimestamp = env.mutationLog[i].time;
      const interval = currentTimestamp - previousTimestamp;

      assert.ok(
        interval <= 1000,
        `Render lag detected at mutation index ${i}: interval of ${interval}ms exceeds maximum permissible 1000ms cadence (DF-LIVENESS-01)`
      );
      previousTimestamp = currentTimestamp;
    }
  });

  it('AC-2: observation window >= 2.5 seconds with no user input registers visible DOM or Canvas pixel mutations (non-static)', async () => {
    await loadMainEntrypoint();

    const app = env.elements.get('app');
    const canvasElement = app.children.find((child) => child.tagName === 'CANVAS');

    let initialPixels = null;
    let initialDomContent = app.innerHTML || app.textContent;

    if (canvasElement) {
      const ctx = canvasElement.getContext('2d');
      initialPixels = Array.from(ctx.getImageData(0, 0, 10, 10).data);
    }

    // 2.5-second (2500ms) observation window without any user input events
    const observationWindowMs = 2500;
    const startMutationIndex = env.mutationLog.length;

    env.advanceTime(observationWindowMs);

    const mutationsDuringWindow = env.mutationLog.slice(startMutationIndex);

    // Verify non-static execution: mutations must have been registered during this idle window
    assert.ok(
      mutationsDuringWindow.length > 0,
      `DF-LIVENESS-01 Failure: No state mutations or frame renders occurred over the ${observationWindowMs}ms observation window. Application appears to be a static painting.`
    );

    // Verify visible difference in DOM text or Canvas buffer
    let stateMutated = false;

    if (canvasElement) {
      const ctx = canvasElement.getContext('2d');
      const currentPixels = Array.from(ctx.getImageData(0, 0, 10, 10).data);
      const canvasDrawCalls = ctx.renderCallCount;

      const pixelsChanged = !initialPixels.every((val, idx) => val === currentPixels[idx]);
      if (pixelsChanged || canvasDrawCalls > 0) {
        stateMutated = true;
      }
    }

    const currentDomContent = app.innerHTML || app.textContent;
    if (currentDomContent !== initialDomContent) {
      stateMutated = true;
    }

    assert.equal(
      stateMutated,
      true,
      'Either Canvas buffer pixels or DOM text content must register visible mutations confirming non-static runtime execution over 2.5s window.'
    );
  });

  it('verifies animation loop schedules via requestAnimationFrame or recurring interval', async () => {
    await loadMainEntrypoint();

    // Verify that at least one RAF or interval is active upon mounting
    const hasActiveRaf = env.rafCallbacks.size > 0;
    const hasActiveInterval = env.timerCallbacks.size > 0;

    assert.ok(
      hasActiveRaf || hasActiveInterval,
      'Mounting src/main.js must register an active requestAnimationFrame loop or tick timer.'
    );

    // Step a single frame and check that the loop re-schedules itself
    env.advanceTime(16.67);

    const persistsActiveRaf = env.rafCallbacks.size > 0;
    const persistsActiveInterval = env.timerCallbacks.size > 0;

    assert.ok(
      persistsActiveRaf || persistsActiveInterval,
      'Active animation loop must continuously re-queue next frame (not terminate after 1 cycle).'
    );
  });
});